File: camount/src/commonMain/kotlin/io/denison/camount/formatter/AmountConfig.kt
package io.denison.camount.formatter

internal class AmountConfig(
  val maximumNotationDigits: Int,
  val decimalSeparator: Char,
  val groupingSeparator: Char,
  val prefix: String,
  val suffix: String,
  val groupingSize: Int,
  val maximumFractionDigits: Int,
) {

  val zero: Char get() = digitChars[0]

  val maximumFormattedSymbols: Int = prefix.length +
    maximumNotationDigits +
    (if (groupingSize == 0) 0 else (maximumNotationDigits - 1) / groupingSize) +
    1 + maximumFractionDigits +
    suffix.length

  fun isDigit(c: Char): Boolean = digitChars.contains(c)
  fun isZero(c: Char): Boolean = zero == c
  fun isInputSeparator(c: Char): Boolean = c == '.' || c == ','
  fun isDecimalSeparator(c: Char): Boolean = decimalSeparator == c
  fun isGroupingSeparator(c: Char): Boolean = groupingSize > 0 && groupingSeparator == c

  fun getDigit(index: Int): Char = digitChars[index]

  private companion object {
    private val digitChars = charArrayOf('0', '1', '2', '3', '4', '5', '6', '7', '8', '9')
  }
}

File: camount-swift/Sources/Camount/Core/AmountConfig.swift
import Foundation

struct AmountConfig: Equatable {
    let maximumNotationDigits: Int
    let decimalSeparator: Character
    let groupingSeparator: Character
    let prefix: String
    let suffix: String
    let groupingSize: Int
    let maximumFractionDigits: Int

    private static let digitChars: [Character] =
        ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9"]

    var zero: Character { Self.digitChars[0] }

    var maximumFormattedSymbols: Int {
        let groupingSeparators =
            groupingSize == 0
            ? 0
            : (maximumNotationDigits - 1) / groupingSize
        return prefix.count + maximumNotationDigits + groupingSeparators + 1 + maximumFractionDigits + suffix.count
    }

    func isDigit(_ c: Character) -> Bool { Self.digitChars.contains(c) }
    func isZero(_ c: Character) -> Bool { zero == c }
    func isInputSeparator(_ c: Character) -> Bool { c == "." || c == "," }
    func isDecimalSeparator(_ c: Character) -> Bool { decimalSeparator == c }
    func isGroupingSeparator(_ c: Character) -> Bool {
        groupingSize > 0 && groupingSeparator == c
    }

    func digit(at index: Int) -> Character { Self.digitChars[index] }
}
